tool/tsh/proxy.go

```diff
@@ -145,94 +168,141 @@ func setupJumpHost(cf *CLIConf, tc *libclient.TeleportClient, sp sshProxyParams)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	sshUserHost := fmt.Sprintf("%s:%s", sp.targetHost, sp.targetPort)
+	if err = sess.RequestSubsystem(proxySubsystemName(sshUserHost, sp.clusterName)); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := proxySession(ctx, sess); err != nil {
 		return trace.Wrap(err)
 	}
 	return nil
 }
 
-func sshProxy(tc *libclient.TeleportClient, params sshProxyParams) error {
-	sshPath, err := getSSHPath()
+func dialSSHProxy(ctx context.Context, tc *libclient.TeleportClient, sp sshProxyParams) (net.Conn, error) {
+	remoteProxyAddr := net.JoinHostPort(sp.proxyHost, sp.proxyPort)
+
+	if !sp.tlsRouting {
+		conn, err := (&net.Dialer{}).DialContext(ctx, "tcp", remoteProxyAddr)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return conn, nil
+	}
+
+	pool, err := tc.LocalAgent().ClientCertPool(sp.clusterName)
 	if err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
+	}
+
+	tlsConfig := &tls.Config{
+		RootCAs:            pool,
+		NextProtos:         []string{string(alpncommon.ProtocolProxySSH)},
+		InsecureSkipVerify: tc.InsecureSkipVerify,
+		ServerName:         sp.proxyHost,
 	}
-	keysDir := profile.FullProfilePath(tc.Config.KeysDir)
-	knownHostsPath := keypaths.KnownHostsPath(keysDir)
 
-	args := []string{
-		"-A",
-		// TODO: remove once we finish deprecating SHA1
-		"-o", fmt.Sprintf("PubkeyAcceptedKeyTypes=+%s", ssh.CertAlgoRSAv01),
-		"-o", fmt.Sprintf("UserKnownHostsFile=%s", knownHostsPath),
-		"-p", params.proxyPort,
-		params.proxyHost,
-		"-s",
-		fmt.Sprintf("proxy:%s:%s@%s", params.targetHost, params.targetPort, params.clusterName),
+	conn, err := (&tls.Dialer{Config: tlsConfig}).DialContext(ctx, "tcp", remoteProxyAddr)
+	if err != nil {
+		return nil, trace.Wrap(err)
 	}
+	return conn, nil
+}
 
-	if tc.HostLogin != "" {
-		args = append([]string{"-l", tc.HostLogin}, args...)
+func proxySubsystemName(userHost, cluster string) string {
+	subsystem := fmt.Sprintf("proxy:%s", userHost)
+	if cluster != "" {
+		return fmt.Sprintf("%s@%s", subsystem, cluster)
 	}
+	return fmt.Sprintf("proxy:%s", userHost)
+}
 
-	log.Debugf("Executing proxy command: %v %v.", sshPath, strings.Join(args, " "))
+func makeSSHClient(ctx context.Context, conn net.Conn, addr string, cfg *ssh.ClientConfig) (*tracessh.Client, error) {
+	cc, chs, reqs, err := tracessh.NewClientConn(ctx, conn, addr, cfg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return tracessh.NewClient(cc, chs, reqs), nil
+}
 
-	child := exec.Command(sshPath, args...)
-	child.Stdin = os.Stdin
-	child.Stdout = os.Stdout
-	child.Stderr = os.Stderr
-	return trace.Wrap(child.Run())
+func proxySession(ctx context.Context, sess *ssh.Session) error {
+	stdout, err := sess.StdoutPipe()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	stdin, err := sess.StdinPipe()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	stderr, err := sess.StderrPipe()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	errC := make(chan error)
```

Comment:
Can fix blocking `rrC <- err` in case of ctx cancel by using buffered channel  ? 
```suggestion
	errC := make(chan error, 3)
```